File: loomc-java/src/main/java/loom/compiler/token/Keywords.java
package loom.compiler.token;

import java.util.ArrayList;
import java.util.Collections;
import java.util.List;

public class Keywords {

	/**
	 * Represents the keywords.
	 * These keywords are reserved and cannot be used as identifiers.
	 * <p>
	 * This is literally just out of laziness, so that I don't have to
	 * type them out every time I need to check for a keyword, and so
	 * I don't have to type out strings every time I need to add a keyword.
	 */
	public enum Keyword {
		AS(false),
		ASYNC(false),
		ATOMIC(false),
		AWAIT(false),
		BREAK(false),
		BOOL(true),
		CHAR(true),
		CLASS(false),
		CONST(false),
		CONSTRUCTOR(false),
		CONTINUE(false),
		DO(false),
		DOUBLE(true),
		ELSE(false),
		ENUM(false),
		F8(true),
		F16(true),
		F32(true),
		F64(true),
		FINAL(false, "fin"),
		FOR(false),
		FUNC(false),
		I8(true),
		I16(true),
		I32(true),
		I64(true),
		IF(false),
		IMPORT(false),
		IN(false),
		NAMESPACE(false),
		NEW(false),
		NULL(false),
		PRIVATE(false, "priv"),
		PROTECTED(false, "prot"),
		PUBLIC(false, "pub"),
		RETURN(false),
		SELF(false),
		STRING(true),
		STRUCT(false),
		TRUE(false),
		FALSE(false),
		TYPE(false),
		VAR(false),
		VOID(true),
		WHILE(false);

		private final boolean isType;
		private final String definedName;

		Keyword (boolean isType, String definedName) {
			this.isType = isType;
			this.definedName = definedName;
		}

		Keyword (boolean isType) {
			this(isType, null);
		}

		public boolean isTypeKeyword () {
			return isType;
		}

		public String getName () {
			return definedName != null ? definedName : name().toLowerCase();
		}

		@Override
		public String toString () {
			return getName();
		}
	}

	private static final List<String> KEYWORDS;

	public static List<String> getKeywords () {
		return Collections.unmodifiableList(KEYWORDS);
	}

	public static boolean isTypeKeyword (String text) {
		for (Keyword kw : Keyword.values()) {
			if (kw.isTypeKeyword() && kw.getName().equals(text)) {
				return true;
			}
		}
		return false;
	}

	static {
		List<String> keywordsList = new ArrayList<>();
		for (Keyword type : Keyword.values()) {
			keywordsList.add(type.getName());
		}
		KEYWORDS = Collections.unmodifiableList(keywordsList);
	}

}

File: loomc/lib/keywords.ml
(** Keywords for the Loom language *)

(** Keyword type with properties *)
type keyword =
  | As
  | Async
  | Atomic
  | Await
  | Break
  | Bool
  | Char
  | Class
  | Const
  | Constructor
  | Continue
  | Do
  | Double
  | Else
  | Enum
  | F8
  | F16
  | F32
  | F64
  | Final
  | For
  | Func
  | I8
  | I16
  | I32
  | I64
  | If
  | Import
  | In
  | Namespace
  | New
  | Null
  | Private
  | Protected
  | Public
  | Return
  | Self
  | String
  | Struct
  | True
  | False
  | Type
  | Var
  | Void
  | While

(** Check if a keyword is a type keyword *)
let is_type_keyword = function
  | Bool | Char | Double | F8 | F16 | F32 | F64 | I8 | I16 | I32 | I64 | String | Void -> true
  | _ -> false

(** Get the string representation of a keyword *)
let to_string = function
  | As -> "as"
  | Async -> "async"
  | Atomic -> "atomic"
  | Await -> "await"
  | Break -> "break"
  | Bool -> "bool"
  | Char -> "char"
  | Class -> "class"
  | Const -> "const"
  | Constructor -> "constructor"
  | Continue -> "continue"
  | Do -> "do"
  | Double -> "double"
  | Else -> "else"
  | Enum -> "enum"
  | F8 -> "f8"
  | F16 -> "f16"
  | F32 -> "f32"
  | F64 -> "f64"
  | Final -> "fin"
  | For -> "for"
  | Func -> "func"
  | I8 -> "i8"
  | I16 -> "i16"
  | I32 -> "i32"
  | I64 -> "i64"
  | If -> "if"
  | Import -> "import"
  | In -> "in"
  | Namespace -> "namespace"
  | New -> "new"
  | Null -> "null"
  | Private -> "priv"
  | Protected -> "prot"
  | Public -> "pub"
  | Return -> "return"
  | Self -> "self"
  | String -> "string"
  | Struct -> "struct"
  | True -> "true"
  | False -> "false"
  | Type -> "type"
  | Var -> "var"
  | Void -> "void"
  | While -> "while"

(** List of all keyword strings *)
let all_keywords =
  [ As; Async; Atomic; Await; Break; Bool; Char; Class; Const; Constructor;
    Continue; Do; Double; Else; Enum; F8; F16; F32; F64; Final; For; Func;
    I8; I16; I32; I64; If; Import; In; Namespace; New; Null; Private;
    Protected; Public; Return; Self; String; Struct; True; False; Type;
    Var; Void; While ]
  |> List.map to_string

(** Check if a string is a type keyword *)
let is_type_keyword_string s =
  List.exists (fun kw -> to_string kw = s && is_type_keyword kw)
    [ Bool; Char; Double; F8; F16; F32; F64; I8; I16; I32; I64; String; Void ]

(** Check if a string is a keyword *)
let is_keyword s =
  List.mem s all_keywords 